File: pkg/dev_compiler/test/codegen_expected/corelib/list_for_each_test.js
dart_library.library('corelib/list_for_each_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__list_for_each_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const collection = dart_sdk.collection;
  const _interceptors = dart_sdk._interceptors;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const list_for_each_test = Object.create(null);
  let JSArrayOfint = () => (JSArrayOfint = dart.constFn(_interceptors.JSArray$(core.int)))();
  let JSArrayOfList = () => (JSArrayOfList = dart.constFn(_interceptors.JSArray$(core.List)))();
  let ListTovoid = () => (ListTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [core.List])))();
  let dynamicTovoid = () => (dynamicTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [dart.dynamic])))();
  let VoidTovoid = () => (VoidTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [])))();
  let dynamicTobool = () => (dynamicTobool = dart.constFn(dart.definiteFunctionType(core.bool, [dart.dynamic])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  list_for_each_test.MyList = class MyList extends collection.ListBase {
    new(list) {
      this.list = list;
    }
    get length() {
      return this.list[dartx.length];
    }
    set length(value) {
      this.list[dartx.length] = value;
    }
    get(index) {
      return this.list[dartx.get](index);
    }
    set(index, val) {
      this.list[dartx.set](index, val);
      return val;
    }
    toString() {
      return dart.toString(this.list);
    }
  };
  dart.addSimpleTypeTests(list_for_each_test.MyList);
  dart.setSignature(list_for_each_test.MyList, {
    constructors: () => ({new: dart.definiteFunctionType(list_for_each_test.MyList, [core.List])}),
    methods: () => ({
      get: dart.definiteFunctionType(dart.dynamic, [core.int]),
      set: dart.definiteFunctionType(dart.void, [core.int, dart.dynamic])
    })
  });
  dart.defineExtensionMembers(list_for_each_test.MyList, [
    'get',
    'set',
    'toString',
    'length',
    'length'
  ]);
  list_for_each_test.testWithoutModification = function(list) {
    let seen = [];
    list[dartx.forEach](dart.bind(seen, dartx.add));
    expect$.Expect.listEquals(list, seen);
  };
  dart.fn(list_for_each_test.testWithoutModification, ListTovoid());
  list_for_each_test.testWithModification = function(list) {
    if (dart.test(list[dartx.isEmpty])) return;
    expect$.Expect.throws(dart.fn(() => list[dartx.forEach](dart.fn(_ => list[dartx.add](0), dynamicTovoid())), VoidTovoid()), dart.fn(e => core.ConcurrentModificationError.is(e), dynamicTobool()));
  };
  dart.fn(list_for_each_test.testWithModification, ListTovoid());
  let const$;
  let const$0;
  let const$1;
  list_for_each_test.main = function() {
    let fixedLengthList = core.List.new(10);
    for (let i = 0; i < 10; i++)
      fixedLengthList[dartx.set](i, i + 1);
    let growableList = core.List.new();
    growableList[dartx.length] = 10;
    for (let i = 0; i < 10; i++)
      growableList[dartx.set](i, i + 1);
    let growableLists = JSArrayOfList().of([[], JSArrayOfint().of([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]), new list_for_each_test.MyList(JSArrayOfint().of([1, 2, 3, 4, 5])), growableList]);
    let fixedLengthLists = JSArrayOfList().of([const$ || (const$ = dart.constList([], dart.dynamic)), fixedLengthList, const$0 || (const$0 = dart.constList([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], core.int)), new list_for_each_test.MyList(const$1 || (const$1 = dart.constList([1, 2], core.int)))]);
    for (let list of growableLists) {
      core.print(list);
      list_for_each_test.testWithoutModification(list);
      list_for_each_test.testWithModification(list);
    }
    for (let list of fixedLengthLists) {
      list_for_each_test.testWithoutModification(list);
    }
  };
  dart.fn(list_for_each_test.main, VoidTodynamic());
  // Exports:
  exports.list_for_each_test = list_for_each_test;
});

File: tests/corelib_2/list_for_each_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import "package:expect/expect.dart";
import "dart:collection";

class MyList extends ListBase {
  List list;
  MyList(this.list);
  get length => list.length;
  set length(value) {
    list.length = value;
  }

  operator [](index) => list[index];
  operator []=(index, val) {
    list[index] = val;
  }

  toString() => list.toString();
}

void testWithoutModification(List list) {
  var seen = [];
  list.forEach(seen.add);

  Expect.listEquals(list, seen);
}

void testWithModification(List list) {
  if (list.isEmpty) return;
  Expect.throws(() => list.forEach((_) => list.add(0)),
      (e) => e is ConcurrentModificationError);
}

main() {
  List fixedLengthList = new List(10);
  for (int i = 0; i < 10; i++) fixedLengthList[i] = i + 1;

  List growableList = new List();
  growableList.length = 10;
  for (int i = 0; i < 10; i++) growableList[i] = i + 1;

  var growableLists = [
    [],
    [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    new MyList([1, 2, 3, 4, 5]),
    growableList,
  ];
  var fixedLengthLists = [
    const [],
    fixedLengthList,
    const [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    new MyList(const [1, 2]),
  ];

  for (var list in growableLists) {
    print(list);
    testWithoutModification(list);
    testWithModification(list);
  }

  for (var list in fixedLengthLists) {
    testWithoutModification(list);
  }
}
